### PR TITLE
ListSpec

### PR DESCRIPTION
**Pull request by kriskowal**
This branch introduces ListSpec, which has some overlap with @jcorbin’s speclist. These need to be consolidated.

- [x] Should bring in #16 to exercise StringSpec and ByteSpec in the tests.
- [x] Remove #29 after it lands

### CODE REVIEW

**Comment by kriskowal** [comment]
r @jcorbin

File: list.js [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+/* eslint max-statements:[1, 30] */
+'use strict';
+
+var bufrw = require('bufrw');
+var TYPE = require('./TYPE');
+var typeNames = require('./names');
+var InvalidSizeError = require('./errors').InvalidSizeError;
```
Comment (jcorbin) [inline list.js]: have we switched over to this way now? I'd vaguely prefer to just bind `var errors` ...

File: list.js [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+var ListTypeIdMismatch = require('./errors').ListTypeIdMismatch;
+
+var LengthResult = bufrw.LengthResult;
+var WriteResult = bufrw.WriteResult;
+var ReadResult = bufrw.ReadResult;
+
+function ListSpec(valueType, annotations) {
```
Comment (jcorbin) [inline list.js]: y u no stash annotations? too much wiggle room for Not Implemented Yet?

File: list.js [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    var self = this;
+    self.rw = new ListRW(valueType, self);
+}
+
+ListSpec.prototype.create = function create() {
+    return [];
+};
+
+ListSpec.prototype.add = function add(list, value) {
```
Comment (jcorbin) [inline list.js]: remind me, these are here for differing behavior in `SetSpec`?  Can we get a comment here to remind future weary travelers?

**Comment by jcorbin** [comment]
Looks okay, but I'll give it another read through after it rebases on its pre-reqs merging.

File: list.js (hunk 1 above)
Comment (kriskowal) [inline list.js]: I left an issue #25 to do a project wide audit to change the style.

File: list.js (hunk 2 above)
Comment (kriskowal) [inline list.js]: Not sure what you’re proposing for stash. Upcoming, SetSpec will be the first illustration of how they are used.

**Comment by jcorbin** [comment]
such rebase, many win, #lgtm

File: list.js (hunk 3 above)
Comment (kriskowal) [inline list.js]: SetSpec upcoming.

**Pull request merged**